File: mpiprog/clang/mpiprog02.c
/* File: mpiprog02.c
 * Name: D.Saravanan
 * Date: 01/12/2023
 * Program with blocking send/receive routines
*/

#include <stdio.h>
#include <mpi.h>

int main(int argc, char **argv) {

    int rank, size;
    int length, mate, mesg;
    char name[MPI_MAX_PROCESSOR_NAME];

    MPI_Status status;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    if (size % 2 != 0) {
		if (rank == 0)
			printf("Quit!. Need an even number of processor, but nproc = %d\n", size);
    }

    else {
		if (rank == 0)
			printf("Number of processor: %d\n", size);

		MPI_Get_processor_name(name, &length);
		printf("rank %d of nproc %d on %s\n", rank, size, name);

		if (rank < size/2) {
			mate = size/2 + rank;
			MPI_Send(&rank, 1, MPI_INT, mate, 1, MPI_COMM_WORLD);
			MPI_Recv(&mesg, 1, MPI_INT, mate, 1, MPI_COMM_WORLD, &status);
		}

		else if (rank >= size/2) {
			mate = rank - size/2;
			MPI_Recv(&mesg, 1, MPI_INT, mate, 1, MPI_COMM_WORLD, &status);
			MPI_Send(&rank, 1, MPI_INT, mate, 1, MPI_COMM_WORLD);
		}

		printf("rank %d is partner with %d\n", rank, mesg);
    }

	MPI_Finalize();

    return 0;
}

File: mpiprog/python/mpiprog02.py
#!/usr/bin/env python
# File: mpiprog02.py
# Name: D.Saravanan
# Date: 01/12/2023

""" Script with blocking send/receive routines """

from mpi4py import MPI


def main():
    """blocking communication"""

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    if size % 2 != 0:
        if rank == 0:
            print(f"Quit! Need an even number of processes, but nproc is {size}.")

    else:
        if rank == 0:
            print(f"Number of processes: {size}")

        name = MPI.Get_processor_name()
        print(f"rank {rank} of nproc {size} on {name}")

        if rank < size / 2:
            mate = size / 2 + rank
            comm.send(rank, dest=mate, tag=11)
            mesg = comm.recv(source=mate, tag=12, status=MPI.Status())

        elif rank >= size / 2:
            mate = rank - size / 2
            mesg = comm.recv(source=mate, tag=11, status=MPI.Status())
            comm.send(rank, dest=mate, tag=12)

        print(f"rank {rank} is partner with {mesg}")


if __name__ == "__main__":
    main()
